gnovm/pkg/gnolang/stacktrace.go

```diff
@@ -0,0 +1,81 @@
+package gnolang
+
+import (
+	"fmt"
+	"strings"
+)
+
+const maxStacktraceSize = 128
+
+type Execution struct {
+	Stmt  Stmt
+	Frame *Frame
+}
+type Stacktrace struct {
+	Executions      []Execution
+	NumFramesElided int
+}
+
+func (s Stacktrace) String() string {
+	var builder strings.Builder
+
+	for i, e := range s.Executions {
+		if s.NumFramesElided > 0 && i == maxStacktraceSize/2 {
+			builder.WriteString(fmt.Sprintf("...%d frame(s) elided...\n", s.NumFramesElided))
+		}
+
+		switch {
+		case e.Stmt == nil:
+			builder.WriteString(fmt.Sprintf("%s()\n", e.Frame.Func))
+			builder.WriteString(fmt.Sprintf("    %s/%s:%d\n", e.Frame.Func.PkgPath, e.Frame.Func.FileName, e.Frame.Func.Source.GetLine()))
+		case e.Frame.Func != nil && e.Frame.Func.IsNative():
+			builder.WriteString(fmt.Sprintf("%s()\n", e.Frame.Func))
+			builder.WriteString(fmt.Sprintf("    %s.%s\n", e.Frame.Func.NativePkg, e.Frame.Func.NativeName))
+		case e.Frame.Func != nil:
+			builder.WriteString(fmt.Sprintf("%s\n", e.Stmt.String()))
```

Comment:
Printing the statement also prints the values which were passed to it, so I think this still has problems if we're trying to print a value which is too big

There should be a way to limit this, at the very least. I'd like if we had a system where it was not recursive (ie. only printing PrimitiveKind, structs without recursion; slices with only length/cap, maps only with length, etc.)

Reply:
done
here: https://github.com/gnolang/gno/pull/2145/commits/c7820cdef701e5ae36af3da35a6d05ade78a98dd, https://github.com/gnolang/gno/pull/2145/commits/5e0cf7e9f6cda64e6aa20b7727958370681f280e